src/rpc/util.cpp

```diff
@@ -635,6 +637,49 @@ UniValue RPCHelpMan::HandleRequest(const JSONRPCRequest& request) const
     return ret;
 }
 
+using CheckFn = void(const RPCArg&);
+static const UniValue* DetailMaybeArg(CheckFn* check, const std::vector<RPCArg>& params, const JSONRPCRequest* req, size_t i)
+{
+    CHECK_NONFATAL(i < params.size());
+    const UniValue& arg{CHECK_NONFATAL(req)->params[i]};
+    const RPCArg& param{params.at(i)};
+    if (check) check(param);
+
+    if (!arg.isNull()) return &arg;
+    if (!std::holds_alternative<RPCArg::Default>(param.m_fallback)) return nullptr;
+    return &std::get<RPCArg::Default>(param.m_fallback);
+}
+
+static void CheckRequiredOrDefault(const RPCArg& param)
+{
+    // Must use `Arg<Type>(i)` to get the argument or its default value.
+    const bool required{
+        std::holds_alternative<RPCArg::Optional>(param.m_fallback) && RPCArg::Optional::NO == std::get<RPCArg::Optional>(param.m_fallback),
+    };
+    CHECK_NONFATAL(required || std::holds_alternative<RPCArg::Default>(param.m_fallback));
+}
+
+#define TMPL_INST(check_param, ret_type, return_code)       \
+    template <>                                             \
+    ret_type RPCHelpMan::ArgValue<ret_type>(size_t i) const \
+    {                                                       \
+        const UniValue* maybe_arg{                          \
+            DetailMaybeArg(check_param, m_args, m_req, i),  \
+        };                                                  \
+        return return_code                                  \
+    }                                                       \
+    void force_semicolon()
```

Comment:
```
../../../src/rpc/util.cpp:671:10: error: redundant redeclaration of ‘void force_semicolon()’ in same scope [-Werror=redundant-decls]
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:675:1: note: in expansion of macro ‘TMPL_INST’
  675 | TMPL_INST(nullptr, std::optional<bool>, maybe_arg ? std::optional{maybe_arg->get_bool()} : std::nullopt;);
      | ^~~~~~~~~
../../../src/rpc/util.cpp:671:10: note: previous declaration of ‘void force_semicolon()’
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:674:1: note: in expansion of macro ‘TMPL_INST’
  674 | TMPL_INST(nullptr, std::optional<double>, maybe_arg ? std::optional{maybe_arg->get_real()} : std::nullopt;);
      | ^~~~~~~~~
../../../src/rpc/util.cpp:671:10: error: redundant redeclaration of ‘void force_semicolon()’ in same scope [-Werror=redundant-decls]
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:676:1: note: in expansion of macro ‘TMPL_INST’
  676 | TMPL_INST(nullptr, const std::string*, maybe_arg ? &maybe_arg->get_str() : nullptr;);
      | ^~~~~~~~~
../../../src/rpc/util.cpp:671:10: note: previous declaration of ‘void force_semicolon()’
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:675:1: note: in expansion of macro ‘TMPL_INST’
  675 | TMPL_INST(nullptr, std::optional<bool>, maybe_arg ? std::optional{maybe_arg->get_bool()} : std::nullopt;);
      | ^~~~~~~~~
../../../src/rpc/util.cpp:671:10: error: redundant redeclaration of ‘void force_semicolon()’ in same scope [-Werror=redundant-decls]
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:679:1: note: in expansion of macro ‘TMPL_INST’
  679 | TMPL_INST(CheckRequiredOrDefault, int, CHECK_NONFATAL(maybe_arg)->getInt<int>(););
      | ^~~~~~~~~
../../../src/rpc/util.cpp:671:10: note: previous declaration of ‘void force_semicolon()’
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:676:1: note: in expansion of macro ‘TMPL_INST’
  676 | TMPL_INST(nullptr, const std::string*, maybe_arg ? &maybe_arg->get_str() : nullptr;);
      | ^~~~~~~~~
../../../src/rpc/util.cpp:671:10: error: redundant redeclaration of ‘void force_semicolon()’ in same scope [-Werror=redundant-decls]
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:680:1: note: in expansion of macro ‘TMPL_INST’
  680 | TMPL_INST(CheckRequiredOrDefault, uint64_t, CHECK_NONFATAL(maybe_arg)->getInt<uint64_t>(););
      | ^~~~~~~~~
../../../src/rpc/util.cpp:671:10: note: previous declaration of ‘void force_semicolon()’
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:679:1: note: in expansion of macro ‘TMPL_INST’
  679 | TMPL_INST(CheckRequiredOrDefault, int, CHECK_NONFATAL(maybe_arg)->getInt<int>(););
      | ^~~~~~~~~
../../../src/rpc/util.cpp:671:10: error: redundant redeclaration of ‘void force_semicolon()’ in same scope [-Werror=redundant-decls]
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:681:1: note: in expansion of macro ‘TMPL_INST’
  681 | TMPL_INST(CheckRequiredOrDefault, const std::string&, CHECK_NONFATAL(maybe_arg)->get_str(););
      | ^~~~~~~~~
../../../src/rpc/util.cpp:671:10: note: previous declaration of ‘void force_semicolon()’
  671 |     void force_semicolon()
      |          ^~~~~~~~~~~~~~~
../../../src/rpc/util.cpp:680:1: note: in expansion of macro ‘TMPL_INST’
  680 | TMPL_INST(CheckRequiredOrDefault, uint64_t, CHECK_NONFATAL(maybe_arg)->getInt<uint64_t>(););
      | ^~~~~~~~~
cc1plus: all warnings being treated as errors
```

Reply:
Seems to work fine without this line.

Reply:
Could drop it and just not have the semi-colons after the macro invocations too.

Could also change it to `void force_semicolon(ret_type dummy)`  to avoid the warning.

clang accepts but ignores `-Wredundant-decls`, clang-tidy defaults to ignoring the warning in macros (yay!), but seems like the current code will be annoying for compiling with gcc.